File: Core/Models/Branch/BranchCompliance.swift
import Foundation

public enum ComplianceStatus: String, Codable, Sendable, Hashable {
    case ok, expiring, expired

    public var labelKey: String { "compliance.\(self)" }
}

public struct BranchCompliance: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var branchID: EntityID
    public var civilDefenceCertNumber: String?
    public var civilDefenceExpiry: Date?
    public var sharjahSportsCouncilRegNumber: String?
    public var sharjahSportsCouncilExpiry: Date?
    public var insurancePolicyNumber: String?
    public var insuranceProvider: String?
    public var insuranceExpiry: Date?
    public var lastHealthSafetyInspectionAt: Date?
    public var lastEmergencyPlanReviewAt: Date?
    public var hasAED: Bool
    public var aedLastServiceAt: Date?
    public var firstAidKitLastCheckedAt: Date?

    public init(
        id: EntityID = UUID(),
        branchID: EntityID,
        civilDefenceCertNumber: String? = nil,
        civilDefenceExpiry: Date? = nil,
        sharjahSportsCouncilRegNumber: String? = nil,
        sharjahSportsCouncilExpiry: Date? = nil,
        insurancePolicyNumber: String? = nil,
        insuranceProvider: String? = nil,
        insuranceExpiry: Date? = nil,
        lastHealthSafetyInspectionAt: Date? = nil,
        lastEmergencyPlanReviewAt: Date? = nil,
        hasAED: Bool = false,
        aedLastServiceAt: Date? = nil,
        firstAidKitLastCheckedAt: Date? = nil
    ) {
        self.id = id
        self.branchID = branchID
        self.civilDefenceCertNumber = civilDefenceCertNumber
        self.civilDefenceExpiry = civilDefenceExpiry
        self.sharjahSportsCouncilRegNumber = sharjahSportsCouncilRegNumber
        self.sharjahSportsCouncilExpiry = sharjahSportsCouncilExpiry
        self.insurancePolicyNumber = insurancePolicyNumber
        self.insuranceProvider = insuranceProvider
        self.insuranceExpiry = insuranceExpiry
        self.lastHealthSafetyInspectionAt = lastHealthSafetyInspectionAt
        self.lastEmergencyPlanReviewAt = lastEmergencyPlanReviewAt
        self.hasAED = hasAED
        self.aedLastServiceAt = aedLastServiceAt
        self.firstAidKitLastCheckedAt = firstAidKitLastCheckedAt
    }

    /// Aggregated severity across all tracked expiries. Any expired/missing
    /// required cert dominates "expiring", which dominates "ok".
    public func status(now: Date = Date()) -> ComplianceStatus {
        let required: [Date?] = [civilDefenceExpiry, sharjahSportsCouncilExpiry, insuranceExpiry]
        var anyExpired = false
        var anyExpiring = false
        for date in required {
            guard let date else { anyExpired = true; continue }
            let days = Calendar(identifier: .gregorian).dateComponents([.day], from: now, to: date).day ?? 0
            if days < 0 { anyExpired = true }
            else if days <= 30 { anyExpiring = true }
        }
        if anyExpired { return .expired }
        if anyExpiring { return .expiring }
        return .ok
    }
}

File: flutter_app/lib/core/models/branch_compliance.dart
import 'entity_id.dart';

/// Port of Core/Models/Branch/BranchCompliance.swift.

enum ComplianceStatus {
  ok,
  expiring,
  expired;

  String get labelKey => 'compliance.$name';

  static ComplianceStatus fromJson(String raw) => ComplianceStatus.values
      .firstWhere((s) => s.name == raw, orElse: () => ComplianceStatus.ok);
}

class BranchCompliance {
  final EntityID id;
  final EntityID branchId;
  final String? civilDefenceCertNumber;
  final DateTime? civilDefenceExpiry;
  final String? sharjahSportsCouncilRegNumber;
  final DateTime? sharjahSportsCouncilExpiry;
  final String? insurancePolicyNumber;
  final String? insuranceProvider;
  final DateTime? insuranceExpiry;
  final DateTime? lastHealthSafetyInspectionAt;
  final DateTime? lastEmergencyPlanReviewAt;
  final bool hasAed;
  final DateTime? aedLastServiceAt;
  final DateTime? firstAidKitLastCheckedAt;

  const BranchCompliance({
    required this.id,
    required this.branchId,
    this.civilDefenceCertNumber,
    this.civilDefenceExpiry,
    this.sharjahSportsCouncilRegNumber,
    this.sharjahSportsCouncilExpiry,
    this.insurancePolicyNumber,
    this.insuranceProvider,
    this.insuranceExpiry,
    this.lastHealthSafetyInspectionAt,
    this.lastEmergencyPlanReviewAt,
    this.hasAed = false,
    this.aedLastServiceAt,
    this.firstAidKitLastCheckedAt,
  });

  /// Aggregated severity across all tracked expiries. Any expired/missing
  /// required cert dominates "expiring", which dominates "ok".
  ComplianceStatus status({DateTime? now}) {
    final checkNow = now ?? DateTime.now();
    final required = [
      civilDefenceExpiry,
      sharjahSportsCouncilExpiry,
      insuranceExpiry,
    ];
    var anyExpired = false;
    var anyExpiring = false;
    for (final date in required) {
      if (date == null) {
        anyExpired = true;
        continue;
      }
      final days = date.difference(checkNow).inDays;
      if (days < 0) {
        anyExpired = true;
      } else if (days <= 30) {
        anyExpiring = true;
      }
    }
    if (anyExpired) return ComplianceStatus.expired;
    if (anyExpiring) return ComplianceStatus.expiring;
    return ComplianceStatus.ok;
  }

  factory BranchCompliance.fromJson(Map<String, dynamic> json) =>
      BranchCompliance(
        id: json['id'] as String,
        branchId: json['branchID'] as String,
        civilDefenceCertNumber: json['civilDefenceCertNumber'] as String?,
        civilDefenceExpiry: json['civilDefenceExpiry'] == null
            ? null
            : DateTime.parse(json['civilDefenceExpiry'] as String),
        sharjahSportsCouncilRegNumber:
            json['sharjahSportsCouncilRegNumber'] as String?,
        sharjahSportsCouncilExpiry: json['sharjahSportsCouncilExpiry'] == null
            ? null
            : DateTime.parse(json['sharjahSportsCouncilExpiry'] as String),
        insurancePolicyNumber: json['insurancePolicyNumber'] as String?,
        insuranceProvider: json['insuranceProvider'] as String?,
        insuranceExpiry: json['insuranceExpiry'] == null
            ? null
            : DateTime.parse(json['insuranceExpiry'] as String),
        lastHealthSafetyInspectionAt:
            json['lastHealthSafetyInspectionAt'] == null
            ? null
            : DateTime.parse(json['lastHealthSafetyInspectionAt'] as String),
        lastEmergencyPlanReviewAt: json['lastEmergencyPlanReviewAt'] == null
            ? null
            : DateTime.parse(json['lastEmergencyPlanReviewAt'] as String),
        hasAed: json['hasAED'] as bool? ?? false,
        aedLastServiceAt: json['aedLastServiceAt'] == null
            ? null
            : DateTime.parse(json['aedLastServiceAt'] as String),
        firstAidKitLastCheckedAt: json['firstAidKitLastCheckedAt'] == null
            ? null
            : DateTime.parse(json['firstAidKitLastCheckedAt'] as String),
      );

  Map<String, dynamic> toJson() => {
    'id': id,
    'branchID': branchId,
    'civilDefenceCertNumber': civilDefenceCertNumber,
    'civilDefenceExpiry': civilDefenceExpiry?.toIso8601String(),
    'sharjahSportsCouncilRegNumber': sharjahSportsCouncilRegNumber,
    'sharjahSportsCouncilExpiry': sharjahSportsCouncilExpiry?.toIso8601String(),
    'insurancePolicyNumber': insurancePolicyNumber,
    'insuranceProvider': insuranceProvider,
    'insuranceExpiry': insuranceExpiry?.toIso8601String(),
    'lastHealthSafetyInspectionAt': lastHealthSafetyInspectionAt
        ?.toIso8601String(),
    'lastEmergencyPlanReviewAt': lastEmergencyPlanReviewAt?.toIso8601String(),
    'hasAED': hasAed,
    'aedLastServiceAt': aedLastServiceAt?.toIso8601String(),
    'firstAidKitLastCheckedAt': firstAidKitLastCheckedAt?.toIso8601String(),
  };
}
